src/hotspot/share/oops/klass.cpp

```diff
@@ -344,11 +344,12 @@ uintx Klass::hash_secondary_supers(Array<Klass*>* secondaries, bool rewrite) {
     return uintx(1) << hash_slot;
   }
 
-  // For performance reasons we don't use a hashed table unless there
-  // are at least two empty slots in it. If there were only one empty
-  // slot it'd take a long time to create the table and the resulting
-  // search would be no faster than linear probing.
-  if (length > SECONDARY_SUPERS_TABLE_SIZE - 2) {
+  // Invariant: _secondary_supers.length >= population_count(_secondary_supers_bitmap)
```

Comment:
It makes sense to assert the invariant in `Klass::set_secondary_supers()` (and, probably, `Klass::restore_unshareable_info()` for a shared klass loaded from CDS archive).